File: autozig-math/zig/dvec4.zig
const std = @import("std");

pub const DVec4 = extern struct {
    x: f64,
    y: f64,
    z: f64,
    w: f64,

    pub fn new(x: f64, y: f64, z: f64, w: f64) DVec4 {
        return .{ .x = x, .y = y, .z = z, .w = w };
    }

    pub fn splat(value: f64) DVec4 {
        return .{ .x = value, .y = value, .z = value, .w = value };
    }

    pub fn dot(self: DVec4, other: DVec4) f64 {
        return self.x * other.x + self.y * other.y + self.z * other.z + self.w * other.w;
    }

    pub fn add(self: DVec4, other: DVec4) DVec4 {
        return .{
            .x = self.x + other.x,
            .y = self.y + other.y,
            .z = self.z + other.z,
            .w = self.w + other.w,
        };
    }

    pub fn sub(self: DVec4, other: DVec4) DVec4 {
        return .{
            .x = self.x - other.x,
            .y = self.y - other.y,
            .z = self.z - other.z,
            .w = self.w - other.w,
        };
    }

    pub fn mul_scalar(self: DVec4, s: f64) DVec4 {
        return .{
            .x = self.x * s,
            .y = self.y * s,
            .z = self.z * s,
            .w = self.w * s,
        };
    }

    pub fn length_squared(self: DVec4) f64 {
        return self.dot(self);
    }

    pub fn length(self: DVec4) f64 {
        return @sqrt(self.length_squared());
    }

    pub fn normalize(self: DVec4) DVec4 {
        const len = self.length();
        if (len == 0) return .{ .x = 0, .y = 0, .z = 0, .w = 0 };
        return self.mul_scalar(1.0 / len);
    }
};

export fn dvec4_new(x: f64, y: f64, z: f64, w: f64) DVec4 {
    return DVec4.new(x, y, z, w);
}

export fn dvec4_dot(self: DVec4, other: DVec4) f64 {
    return self.dot(other);
}

export fn dvec4_add(self: DVec4, other: DVec4) DVec4 {
    return self.add(other);
}

export fn dvec4_sub(self: DVec4, other: DVec4) DVec4 {
    return self.sub(other);
}

export fn dvec4_mul_scalar(self: DVec4, s: f64) DVec4 {
    return self.mul_scalar(s);
}

export fn dvec4_length(self: DVec4) f64 {
    return self.length();
}

export fn dvec4_normalize(self: DVec4) DVec4 {
    return self.normalize();
}

File: autozig-math/src/dvec4.rs
use autozig::include_zig;

#[repr(C)]
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct DVec4 {
    pub x: f64,
    pub y: f64,
    pub z: f64,
    pub w: f64,
}

include_zig!("zig/dvec4.zig", {
    fn dvec4_new(x: f64, y: f64, z: f64, w: f64) -> DVec4;
    fn dvec4_dot(self_: DVec4, other: DVec4) -> f64;
    fn dvec4_add(self_: DVec4, other: DVec4) -> DVec4;
    fn dvec4_sub(self_: DVec4, other: DVec4) -> DVec4;
    fn dvec4_mul_scalar(self_: DVec4, s: f64) -> DVec4;
    fn dvec4_length(self_: DVec4) -> f64;
    fn dvec4_normalize(self_: DVec4) -> DVec4;
});

impl DVec4 {
    pub const ZERO: Self = Self { x: 0.0, y: 0.0, z: 0.0, w: 0.0 };
    pub const ONE: Self = Self { x: 1.0, y: 1.0, z: 1.0, w: 1.0 };
    pub const X: Self = Self { x: 1.0, y: 0.0, z: 0.0, w: 0.0 };
    pub const Y: Self = Self { x: 0.0, y: 1.0, z: 0.0, w: 0.0 };
    pub const Z: Self = Self { x: 0.0, y: 0.0, z: 1.0, w: 0.0 };
    pub const W: Self = Self { x: 0.0, y: 0.0, z: 0.0, w: 1.0 };

    pub fn new(x: f64, y: f64, z: f64, w: f64) -> Self {
        dvec4_new(x, y, z, w)
    }

    pub fn dot(self, other: Self) -> f64 {
        dvec4_dot(self, other)
    }

    pub fn length(self) -> f64 {
        dvec4_length(self)
    }

    pub fn normalize(self) -> Self {
        dvec4_normalize(self)
    }
}

impl std::ops::Add for DVec4 {
    type Output = Self;
    fn add(self, rhs: Self) -> Self::Output {
        dvec4_add(self, rhs)
    }
}

impl std::ops::Sub for DVec4 {
    type Output = Self;
    fn sub(self, rhs: Self) -> Self::Output {
        dvec4_sub(self, rhs)
    }
}

impl std::ops::Mul<f64> for DVec4 {
    type Output = Self;
    fn mul(self, rhs: f64) -> Self::Output {
        dvec4_mul_scalar(self, rhs)
    }
}
